pkg/types/waitgroup.go

```diff
@@ -0,0 +1,49 @@
+package types
+
+import (
+	"sync"
+)
+
+type syncGroupFunc func()
+
+// SyncGroup is essentially a wrapper around sync.WaitGroup, designed for ease of use. You only need to use Add() to
+// add routines and Run() to execute them. When it's time to close or reset, you just need to call WaitAndClear(),
+// which takes care of waiting for all the routines to complete before clearing routine.
+//
+// It eliminates the need for manual management of sync.WaitGroup. Specifically, it highlights that SyncGroup takes
+// care of sync.WaitGroup.Add() and sync.WaitGroup.Done() automatically, reducing the chances of missing these crucial calls.
+type SyncGroup struct {
+	wg sync.WaitGroup
+
+	mu      sync.Mutex
+	sgFuncs []syncGroupFunc
+}
+
+func NewSyncGroup() SyncGroup {
+	return SyncGroup{}
+}
+
+func (w *SyncGroup) WaitAndClear() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.wg.Wait()
+	w.sgFuncs = []syncGroupFunc{}
+}
+
+func (w *SyncGroup) Add(fn syncGroupFunc) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.wg.Add(1)
+	w.sgFuncs = append(w.sgFuncs, fn)
+}
+
+func (w *SyncGroup) Run() {
+	for _, goFn := range w.sgFuncs {
```

Comment:
the mutex is used for w.sgFuncs? so you can do:

lock()
fs := w.sgFuncs
unlock()
for _ , f := fs { ... }

like this?

Reply:
fixed